server/application/application.go

```diff
@@ -760,6 +783,74 @@ func (s *Server) Delete(ctx context.Context, q *application.ApplicationDeleteReq
 	return &application.ApplicationResponse{}, nil
 }
 
+func (s *Server) WatchResourceEvents(q *application.ResourceEventsQuery, ws application.ApplicationService_WatchResourceEventsServer) error {
+	if q.Name == nil || *q.Name == "" {
+		return fmt.Errorf("must provide an application name")
+	}
+	a, appGetErr := s.appclientset.ArgoprojV1alpha1().Applications(s.ns).Get(context.Background(), *q.Name, metav1.GetOptions{})
+	claims := ws.Context().Value("claims")
+	if err := s.enf.EnforceErr(claims, rbacpolicy.ResourceApplications, rbacpolicy.ActionGet, appRBACName(*a)); err != nil {
+		return nil
+	}
+	if appGetErr != nil {
+		return appGetErr
+	}
+
+	// sendIfPermitted is a helper to send the application to the client's streaming channel if the
+	// caller has RBAC privileges permissions to view it
+	sendIfPermitted := func(event v1.Event, eventType watch.EventType) {
+		var group string
+		groupVersion := strings.Split(event.InvolvedObject.APIVersion, "/")
+		if len(groupVersion) > 1 {
+			group = groupVersion[0]
+		}
+		_, err := s.getAppResourceNoEnforce(context.Background(), a, group, event.InvolvedObject.Kind, event.InvolvedObject.Namespace, event.InvolvedObject.Name)
+		if err != nil {
+			return
+		}
+		err = ws.Send(&appv1.ResourceEventWatchEvent{
+			Type:        eventType,
+			Event: event,
+		})
+		if err != nil {
+			return
+		}
+	}
+
+	appEvents := make(chan *appv1.ResourceEventWatchEvent, watchAPIBufferSize)
+	appUnsubscribe := s.appEventBroadcaster.Subscribe(appEvents)
+	defer appUnsubscribe()
+
+	var config *rest.Config
+	config, err := s.getApplicationClusterConfig(context.Background(), a)
+	if err != nil {
+		return err
+	}
+	kubeClientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	eventFactory := informers.NewSharedInformerFactory(kubeClientset, 0)
+	eventInformer := eventFactory.Core().V1().Events().Informer()
```

Comment:
while I remember, we did something similar in Argo Workflows and the events came out of order (i.e. mixed up by time), so we also sorted them in the UI

Reply:
Yep, that's what's currently happening here. I think some performance optimizations might also cause events to be duplicated, so I'll be de-duping in the UI.